torchrl/envs/libs/smac.py

```diff
@@ -0,0 +1,237 @@
+        >>> td = env.reset()
+        >>> td["action"] = env.action_spec.rand()
+        >>> td = env.step(td)
+        >>> print(td)
+        TensorDict(
+            fields={
+                action: Tensor(torch.Size([8, 14]), dtype=torch.int64),
+                done: Tensor(torch.Size([1]), dtype=torch.bool),
+                next: TensorDict(
+                    fields={
+                        observation: Tensor(torch.Size([8, 80]), dtype=torch.float32)},
+                    batch_size=torch.Size([]),
+                    device=cpu,
+                    is_shared=False),
+                observation: Tensor(torch.Size([8, 80]), dtype=torch.float32),
+                reward: Tensor(torch.Size([1]), dtype=torch.float32)},
+            batch_size=torch.Size([]),
+            device=cpu,
+            is_shared=False)
+        >>> print(env.available_envs)
+        ['3m', '8m', '25m', '5m_vs_6m', '8m_vs_9m', ...]
+    """
+
+    git_url = "https://github.com/oxwhirl/smac"
+    available_envs = _get_envs()
+    libname = "smac"
+
+    def __init__(self, env: smac.env.StarCraft2Env = None, **kwargs):
+        if env is not None:
+            kwargs["env"] = env
+        super().__init__(**kwargs)
+
+    def _check_kwargs(self, kwargs: Dict):
+        if "env" not in kwargs:
+            raise TypeError("Could not find environment key 'env' in kwargs.")
+        env = kwargs["env"]
+        if not isinstance(env, (smac.env.StarCraft2Env,)):
+            raise TypeError("env is not of type 'smac.env.StarCraft2Env'.")
+
+    def _build_env(self, env, **kwargs) -> smac.env.StarCraft2Env:
+        # StarCraft2Env must be initialized before _make_specs.
+        env.reset()
+        return env
+
+    def _make_specs(self, env: smac.env.StarCraft2Env) -> None:
+        # Extract specs from definition.
+        self.reward_spec = self._make_reward_spec()
+
+        # Extract specs from instance.
+        # To extract these specs environment must be fully initialized with env.reset().
+        self.input_spec = self._make_input_spec(env)
+        self.observation_spec = self._make_observation_spec(env)
+
+        # TODO: add support for the state.
+        # self.state_spec = self._make_state_spec(env)
+        # self.input_spec["state"] = self._state_spec
+        # self._state_example = self._make_state_example(env)
+
+    def _init_env(self) -> None:
+        pass
+
+    def _make_reward_spec(self) -> TensorSpec:
+        return UnboundedContinuousTensorSpec(device=self.device)
+
+    def _make_input_spec(self, env: smac.env.StarCraft2Env) -> TensorSpec:
+        action_spec = CustomNdOneHotDiscreteTensorSpec(
+            torch.tensor(env.get_avail_actions(), dtype=torch.bool), device=self.device
+        )
+        return CompositeSpec(action=action_spec)
+
+    def _make_observation_spec(self, env: smac.env.StarCraft2Env) -> TensorSpec:
+        info = env.get_env_info()
+        size = torch.Size((info["n_agents"], info["obs_shape"]))
```

Comment:
This Nd specs are doing what we were discussing in https://github.com/pytorch/rl/issues/766#issuecomment-1373338097

Aka we are leaking part of the batch_size (agents) into the specs.

Reply:
If you add the agents to the batch_size, normal specs will likely suffice. But #766 might be helpful for this as well